packages/ipfs-message-port-client/src/client.js

```diff
@@ -0,0 +1,348 @@
+'use strict'
+
+/* eslint-env browser */
+const { decodeError } = require('ipfs-message-port-protocol/src/error')
+
+/**
+ * @template T
+ * @typedef {import('ipfs-message-port-protocol/src/rpc').Remote<T>} Remote
+ */
+
+/**
+ * @template T
+ * @typedef {import('ipfs-message-port-protocol/src/rpc').ProcedureNames<T>} ProcedureNames
+ */
+
+/**
+ * @typedef {Object} QueryOptions
+ * @property {AbortSignal} [signal]
+ * @property {number} [timeout]
+ * @property {Transferable[]} [transfer]
+ */
+
+/**
+ * @template I
+ * @typedef {I & QueryOptions} QueryInput
+ */
+
+/**
+ * Represents server query, encapsulating inputs to the server endpoint and
+ * promise of it's result.
+ *
+ * @template I,O
+ * @class
+ */
+class Query {
+  /**
+   * @param {string} namespace - component namespace on the server.
+   * @param {string} method - remote method this is a query of.
+   * @param {QueryInput<I>} input - query input.
+   */
+  constructor (namespace, method, input) {
+    /** @type {Promise<O>} */
+    this.result = new Promise((resolve, reject) => {
+      this.succeed = resolve
+      this.fail = reject
+      this.signal = input.signal
+      this.input = input
+      this.namespace = namespace
+      this.method = method
+      this.timeout = input.timeout == null ? Infinity : input.timeout
+    })
+  }
+
+  /**
+   * Data that will be structure cloned over message channel.
+   * @returns {Object}
+   */
+  toJSON () {
+    return this.input
+  }
+
+  /**
+   * Data that will be transferred over message channel.
+   * @returns {Transferable[]}
+   */
+  transfer () {
+    return this.input.transfer
+  }
+}
+
+/** @typedef {Transport} ClientTransport */
+
+/**
+ * RPC Transport over `MessagePort` that can execute queries. It takes care of
+ * executing queries by issuing a message with unique ID and fullfilling a
+ * query when corresponding response message is received. It also makes sure
+ * that aborted / timed out queries are calcelled out as needed.
+ *
+ * It is expected that there will be at most one transport for a message port
+ * instance.
+ * @class
+ */
+class Transport {
```

Comment:
I've considered ripping `client.js` / `server.js` pieces into separate library. I think it may make sense but given already too many moving pieces, I would much rather land this and then extract some more general pieces later on.

Please let me know if you agree, if not I'll rip them out.

Reply:
No rush, it can be extracted later.